schema/validate.go

```diff
@@ -6,44 +6,65 @@ import (
 	"os"
 	"path/filepath"
 
+	"github.com/urfave/cli"
 	"github.com/xeipuuv/gojsonschema"
 )
 
-func main() {
-	nargs := len(os.Args[1:])
-	if nargs == 0 || nargs > 2 {
-		fmt.Printf("ERROR: usage is: %s <schema.json> [<document.json>]\n", os.Args[0])
-		os.Exit(1)
-	}
+const usage = `check document with specified schema
+
+   users can use validate in following ways
+
+   1.specify document file as an argument
+      validate --schema <schema.json> <document.json>
+
+   2.pass document content through a pipe
+      cat <document.json> | validate --schema <schema.json>
+
+   3.input document content manually, ended with ctrl+d(or your self-defined EOF keys)
+      validate --schema <schema.json>
+      [INPUT DOCUMENT CONTENT HERE]
+`
 
-	schemaPath, err := filepath.Abs(os.Args[1])
+func validate(context *cli.Context) error {
+	var documentLoader gojsonschema.JSONLoader
+	nargs := context.NArg()
+
+	schemaFile := context.String("schema")
+	if schemaFile == "" {
+		return fmt.Errorf("Error: schema-json file must be specified!")
+	}
+	schemaPath, err := filepath.Abs(schemaFile)
 	if err != nil {
-		fmt.Println(err)
-		os.Exit(1)
+		return fmt.Errorf("Error: invalid schema-file path: %s", err)
 	}
 	schemaLoader := gojsonschema.NewReferenceLoader("file://" + schemaPath)
-	var documentLoader gojsonschema.JSONLoader
 
-	if nargs > 1 {
-		documentPath, err := filepath.Abs(os.Args[2])
+	if nargs == 1 {
+		documentArg := context.Args().Get(0)
+		documentPath, err := filepath.Abs(documentArg)
 		if err != nil {
-			fmt.Println(err)
-			os.Exit(1)
+			return fmt.Errorf("Error: invalid document-file path: %s", err)
 		}
 		documentLoader = gojsonschema.NewReferenceLoader("file://" + documentPath)
-	} else {
+	} else if nargs == 0 {
+		stdinStat, _ := os.Stdin.Stat()
+		if (stdinStat.Mode() & os.ModeCharDevice) != 0 {
+			fmt.Printf("Input your DOCUMENT CONTENT here, ended with ctrl+d or your self-defined EOF keys\n")
```

Comment:
I think this is trying too hard ;).  If silently blocking waiting for stdin is [good enough for `cat`, `grep`, etc.][1], I think we should avoid this [extra cleverness][2] too.

[1]: https://github.com/opencontainers/runtime-spec/pull/552#issuecomment-244140941
[2]: https://github.com/opencontainers/runtime-spec/pull/552#issuecomment-244745149